### PR TITLE
IGNITE-16545 Subscription to revision update in the test configuration framwork

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-16545

### CODE REVIEW

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java [hunk 1]
```diff
@@ -255,27 +292,79 @@ public long notificationCount() {
...
+    }
+
     /**
-     * Checks that instance of the given class can be injected by the extension.
-     *
-     * @param type Field or parameter type.
-     * @return {@code true} if value of the given class can be injected.
+     * Implementation for {@link ConfigurationExtension}.
      */
-    private static boolean supportType(Class<?> type) {
```
Comment (rpuch) [inline modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:278]: The old name (`supportType()`) was good in the sense that it was telling *what was the meaning* of the postfix check. The new name is accurate, but it does not tell you *why* it checks for the postfix.

I would suggest to go back to the old approach (but change the name as this class now injects not only configurations, so my suggestion is to rename it to `supportsAsConfigurationType()`).

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java [hunk 2]
```diff
@@ -255,27 +292,79 @@ public long notificationCount() {
...
+    private static List<Field> getInjectRevisionListenerHolderFields(Class<?> testClass) {
+        return AnnotationSupport.findAnnotatedFields(
+                testClass,
+                InjectRevisionListenerHolder.class,
+                field -> isRevisionListenerHolder(field.getType()),
+                HierarchyTraversalMode.TOP_DOWN
+        );
+    }
+
+    private static boolean isNameEndsWithConfiguration(Class<?> type) {
```
Comment (rpuch) [inline modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:313]: `doesNameEndWithConfiguration()` seems to be more gramatically correct. Another option is `isNameEndingWithConfiguration()`.

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java [hunk 3]
```diff
@@ -255,27 +292,79 @@ public long notificationCount() {
...
+
+        /** {@inheritDoc} */
+        @Override
+        public void stopListenUpdateStorageRevision(ConfigurationStorageRevisionListener listener) {
+            listeners.removeListener(listener);
+        }
+
+        private Collection<CompletableFuture<?>> notifyStorageRevisionListeners(long storageRevision, long notificationNumber) {
+            // Lazy init.
+            List<CompletableFuture<?>> futures = null;
```
Comment (rpuch) [inline modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:345]: Does a test class need to try avoiding an allocation? It seems that the performance is not important here, but this trick complicates the code a little.

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java [hunk 4]
```diff
@@ -102,28 +111,42 @@ public void beforeEach(ExtensionContext context) throws Exception {
...
     /** {@inheritDoc} */
     @Override
     public boolean supportsParameter(
-            ParameterContext parameterContext, ExtensionContext extensionContext
+            ParameterContext parameterContext,
+            ExtensionContext extensionContext
     ) throws ParameterResolutionException {
-        return parameterContext.isAnnotated(InjectConfiguration.class)
-                && supportType(parameterContext.getParameter().getType());
+        Class<?> parametrType = parameterContext.getParameter().getType();
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:146]: Fix it.

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java [hunk 5]
```diff
@@ -216,13 +246,20 @@ private static Object cfgValue(
                     ConfigurationUtil.dropNulls(copy);
 
                     if (superRootRef.compareAndSet(sr, copy)) {
-                        Collection<CompletableFuture<?>> futures = notifyListeners(
+                        long storageRevision = revisionListenerHolder.storageRev.incrementAndGet();
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:249]: **StorageRevisionListenerHolderImpl** is **private** and does not require more encapsulation as the code is not complicated.

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java (hunk 3 above)
Comment (tkalkirill) [reply modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:345]: Agreed, fixed it.

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java (hunk 1 above)
Comment (tkalkirill) [reply modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:278]: Name you suggested makes sense.

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java (hunk 2 above)
Comment (tkalkirill) [reply modules/configuration/src/test/java/org/apache/ignite/internal/configuration/testframework/ConfigurationExtension.java:313]: Changed the name.

**Comment by tkalkirill** [comment]
@rpuch Can you review again?

**Review by rpuch — APPROVED** [review]

**Pull request merged**